numpy/_core/src/multiarray/methods.c

```diff
@@ -1312,6 +1308,35 @@ array_sort(PyArrayObject *self,
     if (val < 0) {
         return NULL;
     }
+    if (descending) {
+        int ndim = PyArray_NDIM(self);
+        int normalized_axis = axis;
+
+        if (check_and_adjust_axis(&normalized_axis, ndim) < 0) {
+            return NULL;
+        }
+
+        PyObject *indices = PyTuple_New(ndim);
+
+        for (int i=0; i<ndim; i++) {
+            PyObject *slice;
+            if (i == normalized_axis) {
+                slice = PySlice_New(Py_None, Py_None, PyLong_FromLong(-1));
+            } else {
+                slice = PySlice_New(Py_None, Py_None, Py_None);
+            }
+            PyTuple_SetItem(indices, i, slice);
+        }
+
+        PyArrayObject* result = (PyArrayObject *)array_subscript(self, indices);
+        if (result < 0) {
+            return NULL;
+        }
+        if (PyArray_CopyInto(self, result) < 0) {
```

Comment:
I think here you do not want to do a copy; rather, you need to copy the strides and data pointer from `result`. I think you may be better off just to generate those directly instead of going through `array_subscript`.

I cannot help but think, though, that the better solution is to expose sorting by descending order... It may be OK if not all methods support that (as long as the default does...).

Reply:
> I think here you do not want to do a copy; rather, you need to copy the strides and data pointer from `result`. I think you may be better off just to generate those directly instead of going through `array_subscript`.

That sounds right to me - definitely the first part, and probably also the second part. `array_subscript` usage is pretty rare.

> I cannot help but think, though, that the better solution is to expose sorting by descending order... It may be OK if not all methods support that (as long as the default does...).

That sounds like both a lot more work, potentially duplicated implementations, and less functionality. Doing the regular sort and inverting the result by viewing the sorted array with inverted strides seems like the natural implementation to me, is cheap performance-wise, and easy to implement in a single place.

Reply:
Agreed it is easier, though as @seberg noted one has to be a bit careful for `stable` (arg)sort, since the order of identical elements has to remain stable: see https://github.com/numpy/numpy/pull/25437#discussion_r1444031937 and what follows there (testing this will probably need `object` arrays for regular sort). Though really the test cases are likely the most work, and those have to be written regardless...